### ISSUE TITLE
A couple questions.

### ISSUE THREAD

**Issue by nickcharles** [fredsterboard/vector_library#1]
Hey Fred, saw your post on HH and thought this was pretty cool. You've done this in an interesting way by using macros to dynamically generate code based on what type the user is interested in using in their vector. Have you thought about writing it in such a way that the vector is an array of void pointers that point to the contained objects and cast appropriately when needed? I think writing it this way would give you more idiomatic code, and if I remember correctly macros don't get much help from the compiler.

Also, I'd consider moving to an MIT license, as I believe the GPL requires that any software using this library would be required to also use the GPL. From what I understand, that would force developers to make it open-source.

Looking forward to hearing your thoughts.

**Comment by fredsterboard** [comment]
Thanks for asking! 
To my knowledge, all this macro does is replace the text within your code with a new struct definition and its corresponding function pointers after the preprocessor is called. I see no reason as to why these macros would give the compiler any trouble, as it would just look like any other code at compile time.

As for the reasons I didn't just make the elements void \* in the array, I have one syntactic reason and one performance reason.

syntactic reason: I didn't want to make it necessary to have to typecast and dereference every time you used an element of the vector. I also didn't want to make it necessary to allocate memory to the element you're working with before passing it into the vector. It just makes for more to manage on the implementation side and the user side. I also wanted to make it as close as possible to resembling an object in other languages without completely losing sense of what's going on.

performance reason: If all your elements are just pointers, then the data that all your elements hold are going to be in varying blocks of memory, similar to how the data of linked lists are stored in varying places in memory. This is problematic if you have a gigantic vector as all the actual data that you care about is not in the same contingent block of memory. All your ints (for example) would be in several places in memory (as all you hold are pointers to them), making it less likely for them to be cached together. I have seen situations where having all your data in the same place increases performance by a factor of 50x due to cache optimization. This is one reason why vectors can perform much better than a linked list, even if you were to only use the vector as a linked list.

As for the license, I guess I'll change it to an MIT License. Thanks!

**Comment by nickcharles** [comment]
Yea now that I think about it you're absolutely right, it's just a replacement.

Those are the two reasons I can think of that are negatives against my recommendation. As long as you've considered them I'd say you're good to go. Though I do stand by my statement that it doesn't feel very idiomatic, for example with `vector(int) x` actually declaring a pointer to a vector of ints.

Great work on this, you may have inspired me to write a version of my own using void pointers.

**Comment by fredsterboard** [comment]
Yeah, it's not the most intuitive thing ever, hence the documentation to support what's going on with the macros. And awesome! I definitely would like to see that.
